compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -3013,6 +3013,28 @@ public void visit(BLangTypeInit cIExpr) {
                 if (!cIExpr.initInvocation.argExprs.isEmpty()) {
                     BLangExpression iteratorExpr = cIExpr.initInvocation.argExprs.get(0);
                     BType constructType = checkExpr(iteratorExpr, env, symTable.noType);
+                    BType closeableIteratorType = symTable.langQueryModuleSymbol.scope
+                            .lookup(names.fromString(BLangCompilerConstants.ABSTRACT_CLOSEABLE_ITERATOR_CLASS))
+                            .symbol.type;
+                    BType iteratorType = symTable.langQueryModuleSymbol.scope
+                            .lookup(names.fromString(BLangCompilerConstants.ABSTRACT_ITERATOR_CLASS)).symbol.type;
+                    BAttachedFunction closeFunc = types.getAttachedFuncFromObject((BObjectType) constructType,
+                            BLangCompilerConstants.CLOSE_FUNC);
```

Comment:
Can't we look up the two functions, `next` and `close`, and if not null check if the `isolated` flag is on? That way we can give a more descriptive error, avoid looking up the iterators for each iteration, and also continue the rest of the analysis for the object?